lib/auth/join_oracle.go

```diff
@@ -18,20 +18,224 @@ package auth
+	}
+
+	// Check challenge.
+	if headers.Get(oracle.ChallengeHeader) != challenge {
+		return trace.BadParameter("challenge does not match")
+	}
+
+	return nil
+}
+
+func checkOracleAllowRules(claims oracle.Claims, token string, allowRules []*types.ProvisionTokenSpecV2Oracle_Rule) error {
+	for _, rule := range allowRules {
+		if rule.Tenancy != claims.TenancyID {
+			continue
+		}
+		if len(rule.ParentCompartments) != 0 && !slices.Contains(rule.ParentCompartments, claims.CompartmentID) {
+			continue
+		}
+		if len(rule.Regions) != 0 && !slices.ContainsFunc(rule.Regions, func(region string) bool {
+			return oracle.ParseRegion(region) == claims.Region()
+		}) {
+			continue
+		}
+		return nil
+	}
+	return trace.AccessDenied("instance %v did not match any allow rules in token %v", claims.InstanceID, token)
+}
+
+func formatHeaderFromMap(m map[string]string) http.Header {
+	header := make(http.Header, len(m))
+	for k, v := range m {
+		header.Set(k, v)
+	}
+	return header
+}
+
+func getRegionFromHost(host string) (string, error) {
+	regionEndpoint, foundAuth := strings.CutPrefix(host, "auth.")
+	rawRegion, foundOracleCloud := strings.CutSuffix(regionEndpoint, ".oraclecloud.com")
+	if !foundAuth || !foundOracleCloud {
+		return "", trace.BadParameter("expected host auth.{region}.oraclecloud.com, got %s", host)
+	}
+	region := oracle.ParseRegion(rawRegion)
+	if region == "" {
+		return "", trace.BadParameter("missing or invalid region")
+	}
+	return region, nil
+}
+
+func (a *Server) checkOracleRequest(
+	ctx context.Context,
+	challenge string,
+	tokenReq *types.RegisterUsingTokenRequest,
+	oracleReq *proto.OracleSignedRequest,
+	fetchClaims oracleClaimsFetcher,
+) (oracle.Claims, error) {
+	// Check shared token parameters.
+	provisionToken, err := a.GetToken(ctx, tokenReq.Token)
+	if err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+	if provisionToken.GetJoinMethod() != types.JoinMethodOracle {
+		return oracle.Claims{}, trace.AccessDenied("this token does not support the Oracle join method")
+	}
+
+	// Check signed headers.
+	outerHeaders := formatHeaderFromMap(oracleReq.Headers)
+	innerHeaders := formatHeaderFromMap(oracleReq.PayloadHeaders)
+	if err := checkHeaders(outerHeaders, challenge, a.GetClock()); err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+	if err := checkHeaders(innerHeaders, challenge, a.GetClock()); err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+
+	// Authenticate request with Oracle.
+	region, err := getRegionFromHost(outerHeaders.Get("host"))
+	if err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+	authReq, err := oracle.CreateRequestFromHeaders(region, innerHeaders, outerHeaders)
+	if err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+	claims, err := fetchClaims(ctx, authReq)
+	if err != nil {
+		return oracle.Claims{}, trace.Wrap(err)
+	}
+
+	// Check allow rules.
+	token, ok := provisionToken.(*types.ProvisionTokenV2)
```

Comment:
What if we introduce V3? Perhaps using `provisionToken.GetVersion()` and asserting a minimum version here would be more future-proof?

Reply:
We'd still need to cast to v2 to get the v2 spec.